### PR TITLE
Fix #42 - clarify meaning of scope of scripting variables for actions

### LINKED ISSUE #42: Clarification required for declaration of scripting variables for actions
There are some grey areas in the JSP specification regarding when and how scripting variables associated with actions are declared that need to be clarified.

Consider the following from JSP.9.4.4
```
<x:tag>  
  foo
</x:tag>
```
Equivalent Text:
```
declare AT_BEGIN variables
{
  declare NESTED variables
  transformation of foo
}
declare AT_END variables
```

If <x:tag> declares an AT_BEGIN scripting variable myVar of type String then the equivalent text becomes (ignoring the required variable synchronization to make the example clearer)
```
String myVar = null;
{
  transformation of foo
}
```

If our original code is extended to:
```
<x:tag>  
  foo
</x:tag>
<x:tag>  
  foo
</x:tag>
```

then the equivalent text becomes

```
String myVar = null;
{
  transformation of foo
}
String myVar = null;
{
  transformation of foo
}
```

Clearly this is problematic since it is illegal to declare the same variable multiple times in the same scope.

There are a couple of ways to handle this. The first is to insert an outer code block around the code for each tag. However, that makes little sense since it makes the declaration of AT_END variables pointless as the following example shows. There is no point declaring the AT_END variables in this case since they instantly go out of scope.

```
{
  declare AT_BEGIN variables
  {
    declare NESTED variables
    transformation of foo
  }
  declare AT_END variables
}
{
  declare AT_BEGIN variables
  {
    declare NESTED variables
    transformation of foo
  }
  declare AT_END variables
}
```

Another alternative is to pre-process the tags and to ensure that duplicate scripting variables are not defined. The equivalent text for multiple tag instances then becomes:

```
String myVar = null;
{
  transformation of foo
}
{
  transformation of foo
}
```

The pre-processing approach works in most cases but it can break if scriptlets are present. Consider the following original text:

```
<% try { %>
  <x:tag>  
    foo
  </x:tag>
<% } catch (IllegalArgumentException iae) { %>
  <x:tag>  
    foo
  </x:tag>
<% } %>
```
With pre-processing this becomes:
```
try {
  String myVar = null;
  {
    transformation of foo
  }
} catch (IllegalArgumentException iae) {
  {
    transformation of foo
  }
}
```

which will break if the transformation of foo refers to myVar (which it almost certainly will) because myVar is now out of scope for the second instance of the tag.

I do not believe it is reasonable or practical for JSP containers to parse scriplets surrounding actions to determine the scope of any declared scripting variables and then adjust the declaration of those scripting variables accordingly.

One potential solution is to move the declaration of all scripting variables for actions (AT_BEGIN, NESTED and AT_END) to immediately after the declaration of the implicit scripting variables so the scripting variables for actions are always declared once only and are always in scope during the execution of the page.

It is necessary for the JSP specification to clarify how scripting variables declared by actions should be handled. It should be noted that there are variations on this problem such as when tags are nested.

For background on the Tomcat bug report that prompted this issue see:
[https://issues.apache.org/bugzilla/show_bug.cgi?id=56516](https://issues.apache.org/bugzilla/show_bug.cgi?id=56516)
Also see:
[https://issues.apache.org/bugzilla/show_bug.cgi?id=42390](https://issues.apache.org/bugzilla/show_bug.cgi?id=42390)
[https://issues.apache.org/bugzilla/show_bug.cgi?id=48616](https://issues.apache.org/bugzilla/show_bug.cgi?id=48616)

### PR DESCRIPTION
**Pull request by markt-asf**

### CODE REVIEW

File: api/src/main/java/jakarta/servlet/jsp/tagext/VariableInfo.java [hunk 1]
```diff
@@ -72,12 +76,12 @@
...
- * it.
- * <li>AT_BEGIN, if the scripting variable is available from the start tag of the action that defines it until the end
- * of the scope.
- * <li>AT_END, if the scripting variable is available after the end tag of the action that defines it until the end of
- * the scope.
+ * <li>NESTED, if the scripting variable is synchronized with the calling page between the start tag and the end tag of
+ * the action that defines it.
+ * <li>AT_BEGIN, if the scripting variable is synchronized with the calling page from the start tag of the action that
+ * defines it until the end of the scope.
+ * <li>AT_END, if the scripting variable is synchronized with the calling page after the end tag of the action that
```
Comment (joakime) [inline api/src/main/java/jakarta/servlet/jsp/tagext/VariableInfo.java:83]: The word "synchronized" in this context seems odd.
Can you explain what you mean here?

**Review by joakime — COMMENTED** [review]
Can you clarify this?

File: api/src/main/java/jakarta/servlet/jsp/tagext/VariableInfo.java (hunk 1 above)
Comment (markt-asf) [reply api/src/main/java/jakarta/servlet/jsp/tagext/VariableInfo.java:83]: That wording based on section 8.9.1 of the spec document.
What I was trying to do was make clear that "scope" in this case does not mean "scope in the sense of where is this variable declaration visible", but "how and when the scripting variable is kept in sync with the calling page".

Comment (markt-asf) [reply api/src/main/java/jakarta/servlet/jsp/tagext/VariableInfo.java:83]: Is that explanation sufficient or are you looking for some changes in wording?

**Review by joakime — APPROVED** [review]

**Pull request merged**